Containerized subpath

jsafrane opened the pull request:
**What this PR does / why we need it**:
Containerized kubelet needs a different implementation of `PrepareSafeSubpath` than kubelet running directly on the host.

On the host we safely open the subpath and then bind-mount `/proc/<pidof kubelet>/fd/<descriptor of opened subpath>`.

With kubelet running in a container, `/proc/xxx/fd/yy` on the host contains path that works only inside the container, i.e. `/rootfs/path/to/subpath` and thus any bind-mount on the host fails.

Solution:
- safely open the subpath and gets its device ID and inode number
- blindly bind-mount the subpath to `/var/lib/kubelet/pods/<uid>/volume-subpaths/<name of container>/<id of mount>`. This is potentially unsafe, because user can change the subpath source to a link to a bad place (say `/run/docker.sock`) just before the bind-mount.
- get device ID and inode number of the destination. Typical users can't modify this file, as it lies on /var/lib/kubelet on the host.
- compare these device IDs and inode numbers.

**Which issue(s) this PR fixes**
Fixes #61456

**Special notes for your reviewer**:

The PR contains some refactoring of `doBindSubPath` to extract the common code. New `doNsEnterBindSubPath` is added for the nsenter related parts.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

jsafrane commented:
/test pull-kubernetes-cross 
Calling explicitly to check Windows mounter

jsafrane commented:
/sig storage
@msau42 PTAL
@cofyc, you've played with nsenter recently, would you mind taking a look?

BTW, we need real unit test for whole nsenter mounter.

msau42 commented:
/assign

msau42 commented on pkg/util/mount/mount.go:
```diff
@@ -112,6 +112,11 @@ type Interface interface {
 	// It returns the same path as it gets unless kubelet runs in a container.
 	// Then it returns "/rootfs/" + <path>
 	KubeletPath(path string) string
+	// EvalSymlinks returns the path name after the evaluation of any symbolic
```
returns the path name on the host

msau42 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -175,7 +174,8 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 			}
 
-			fileinfo, err := os.Lstat(hostPath)
+			kubeletHostPath := mounter.KubeletPath(hostPath)
```
Instead of exposing KubeletPath as an interface, would it be better to expose Lstat as an interface, and have EvalSymlinks return container path?

jsafrane replied:
Actually the code here is wrong. It resolves any symlinks in the `hostPath` in kubelet, which is not possible.

And that's Lstat is not easy - we need to resolve the symlinks on the host and then do `os.Lstat("/rootfs/" + resolved_path)` in kubelet

jsafrane replied:
Fixed by some refactoring.

jsafrane commented:
Pushed 2nd version.

kubelet_pods.go now passes all paths as paths on the host to mounter and mounter translates them by itself when it's needed.

I changed semantics of `SafeMakeDir` 
from `SafeMakeDir(*fullPath* string, base string, perm os.FileMode)`
to `SafeMakeDir(*subdirectoryInBase* string, base string, perm os.FileMode)`
to simplify its implementation a bit.

@andyzhangx, I touched the Windows part a bit, please review.

jsafrane commented:
@kubernetes/sig-storage-pr-reviews, any review is highly appreciated, we don't want more regressions in subpath!

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -387,3 +367,85 @@ func (mounter *NsenterMounter) evalHostSymlinks(path string, mustExist bool) (st
+	}
+	defer syscall.Close(fd)
+
+	glog.Infof("JSAF before prepare")
```
Cleanup

jsafrane replied:
Cleaned up

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -282,7 +274,15 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+		return false, err
+	}
+	kubeletPath := mounter.getKubeletPath(evaluatedPath)
+	_, err = os.Lstat(kubeletPath)
```
what was wrong with mounter.Lstat()?

jsafrane replied:
Reworked `mounter.Lstat()` to return `NotExist` when the target does not exist (it returned some indistinguishable error before).

msau42 commented on hack/local-up-cluster.sh:
```diff
@@ -800,7 +800,7 @@ function start_kubelet {
 
       docker run --rm --name kubelet \
-        --volume=/:/rootfs:ro,rslave \
+        --volume=/:/rootfs:rw,rslave \
```
does this mean existing deployments will break if they upgrade to this fix and don't change this setting?

Previously, safemkdir still worked for non-hostpath types because it uses /var/lib/kubelet/.... instead of /rootfs/var/lib/kubelet

jsafrane replied:
That's right, it will break.

I could pass `podDir` into `SafeMakeDir()` as a new parameter. `NsenterMounter.SafeMakeDir` could then check if the directory to create is subdir of `podDir` and add `/rootfs` only when it's not.

Any better ideas?

jsafrane replied:
I added a new argument to `NewNsenterMounter` so the mounter knows what's location of `/var/lib/kubelet` and can do a shortcut in `SafeMakeDir`. That was the easiest way I could find out. PTAL.

jsafrane replied:
And I removed changes in hack/local-up-cluster.sh

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -281,49 +273,179 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+		return fmt.Errorf("error resolving symlinks in %s: %s", base, err)
 	}
-	return strconv.Atoi(string(matches[1]))
+	kubeletBase := mounter.getKubeletPath(evaluatedBase)
```
Do the paths need to be cleaned?

jsafrane replied:
Added to `doSafeMakeDir`, just to be sure.

msau42 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -687,102 +684,134 @@ func doMakeRShared(path string, mountInfoFilename string) error {
+}
 
-	// Check it's not already bind-mounted
+// prepareSubpathTarget creates target for bind-mount of subpath. It returns
```
Add comment this is also used by nsenter

jsafrane replied:
Comment added

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -281,49 +273,179 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+	defer syscall.Close(fd)
+
+	glog.Infof("JSAF before prepare")
+	alreadyMounted, bindPathTarget, err := prepareSubpathTarget(mounter, subpath)
```
should this be done first?

jsafrane replied:
IMO it does not really matter if we do `safeOpenSubPath` or `prepareSubpathTarget` first, I wanted the `defer` that cleans up volumes to affect as little code as possible.

msau42 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -687,102 +684,134 @@ func doMakeRShared(path string, mountInfoFilename string) error {
 	}
 	defer syscall.Close(fd)
 
+	alreadyMounted, bindPathTarget, err := prepareSubpathTarget(mounter, subpath)
```
Check this first?

cofyc commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -281,49 +273,179 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+// path: /mnt/volume/non/existing/directory. /mnt/volume exists and
+// non/existing/directory does not exist. It resolves symlinks in /mnt/volume
+// to say /mnt/foo and returns /mnt/foo/non/existing/directory.
+func (mounter *NsenterMounter) evalHostSymlinks(path string, mustExist bool) (string, error) {
```
I had added a similar function in https://github.com/kubernetes/kubernetes/pull/62903/files#diff-f6ffeca7a3b942208d644eaa4ec99642R127 as you and @msau42 suggested. How about merging https://github.com/kubernetes/kubernetes/pull/62903 first, then use `Nsenter.HostPath` to get the path name on the host after evaluating symlinks on the host (and adding `mustExist` parameter)?

jsafrane commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -33,7 +33,7 @@ import (
 const (
 	// hostProcMountsPath is the default mount path for rootfs
-	hostProcMountsPath = "/rootfs/proc/1/mounts"
+	HostProcMountsPath = "/rootfs/proc/1/mounts"
```
It does not need to be public if we introduce `Nsenter.HostPath()` as in https://github.com/kubernetes/kubernetes/pull/62903/files#diff-f6ffeca7a3b942208d644eaa4ec99642R129

msau42 replied:
I only see it being used in this file.

jsafrane commented on pkg/util/mount/mount_windows.go:
```diff
@@ -436,8 +436,9 @@ func getAllParentLinks(path string) ([]string, error) {
-func (mounter *Mounter) SafeMakeDir(pathname string, base string, perm os.FileMode) error {
-	return doSafeMakeDir(pathname, base, perm)
+func (mounter *Mounter) SafeMakeDir(subdir string, base string, perm os.FileMode) error {
+	fullPath := filepath.Join(base, subdir)
```
good catch, fixed

dims commented:
looks like we may break other platforms based on what pull-kubernetes-typecheck is saying ..

```
ERROR(darwin/386,darwin/amd64,windows/386,windows/amd64) cmd/kubelet/app/server.go:342:42: too many arguments
```

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -299,18 +305,30 @@ func (mounter *NsenterMounter) PrepareSafeSubpath(subPath Subpath) (newHostPath
-	kubeletBase := mounter.getKubeletPath(evaluatedBase)
 
+	// Base is somewhere on the host's filesystem. Add /rootfs and try to make
+	// the directory there.
```
Add comment that this requires /rootfs to be writeable.

jsafrane replied:
added

msau42 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -687,102 +684,137 @@ func doMakeRShared(path string, mountInfoFilename string) error {
+		return "", fmt.Errorf("error resolving symlinks in %q: %v", subpath.Path, err)
+	}
+	glog.V(5).Infof("doBindSubPath %q (%q) for volumepath %q", subpath.Path, newPath, subpath.VolumePath)
+	subpath.VolumePath = filepath.Clean(newVolumePath)
```
Clean is not needed here since it's done in doSafeOpen()?

jsafrane replied:
fixed

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -281,49 +279,182 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+	evaluatedSubdirPath = filepath.Clean(evaluatedSubdirPath)
+
+	rootDir := filepath.Clean(mounter.rootDir)
+	if pathWithinBase(evaluatedBase, rootDir) {
```
unit test for this?

msau42 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -175,29 +174,27 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 			}
 			perm := fileinfo.Mode()
 
-			volumePath, err := filepath.EvalSymlinks(hostPath)
```
Does windows PrepareSafeSupath need to evalsymlinks too?

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -281,49 +279,182 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+		return "", fmt.Errorf("error mounting %s: %s", hostSubpath, err)
+	}
+
+	// Check that the bind-mount target is the same inode and device as the
```
Is it possible to unit test this section?

andyzhangx commented:
@jsafrane you may also build on Windows and run windows unit test on pkg\util\mount, someone is working on making windows unit test as CI test, while currently we need to run it locally.

```
KUBE_BUILD_PLATFORMS=windows/amd64 make WHAT=cmd/kubelet

c:\Go\src\k8s.io\kubernetes\pkg\util\mount>go test -v
# k8s.io/kubernetes/pkg/util/mount
.\mount.go:147: cannot use Mounter literal (type *Mounter) as type Interface in assignment:
        *Mounter does not implement Interface (missing GetMode method)
.\mount_windows.go:48: cannot use Mounter literal (type *Mounter) as type Interface in return argument:
        *Mounter does not implement Interface (missing GetMode method)
.\mount_windows.go:139: cannot use mounter (type *Mounter) as type Interface in argument to IsNotMountPoint:
        *Mounter does not implement Interface (missing GetMode method)
.\mount_windows.go:158: cannot use mounter (type *Mounter) as type Interface in argument to getDeviceNameFromMount:
        *Mounter does not implement Interface (missing GetMode method)
.\mount_windows.go:444: cannot use mounter (type *Mounter) as type Interface in argument to getMountRefsByDev:
        *Mounter does not implement Interface (missing GetMode method)
.\mount_windows.go:453: undefined: fakeMounter
FAIL    k8s.io/kubernetes/pkg/util/mount [build failed]
```

andyzhangx reviewed (changes requested):
@jsafrane and the unit test `TestDoSafeMakeDir` in mount_windows_test.go should also change since this PR use `subdir` instead of full path

jsafrane commented:
/test pull-kubernetes-cross

jsafrane commented:
/test pull-kubernetes-local-e2e-containerized

jsafrane commented:
> the unit test TestDoSafeMakeDir in mount_windows_test.go should also change since this PR use subdir instead of full path

@andyzhangx, `SafeMakeDir` joins the paths together and `doSafeMakeDir` did not change at all.

jsafrane commented:
/retest

jsafrane commented:
Added bunch of unit tests for Nsenter and NsenterMounter.

I noticed that `realpath -m` does not report any error, i.e. on a symlink loop or "access denied" errors. It may happen that a symlink that can't be resolved on the host can be resolved in container. These two methods are affected:

* ExistsPath() - it may report that a file exists when it's not.
* `SafeMakeDir` will fail correctly.

I leave it as it is for now, @cofyc, do you still have your EvalSymlinks implementation that just reads files in /rootfs and resolves symlinks there?

jsafrane commented:
/test pull-kubernetes-cross

jsafrane commented:
/retest

jsafrane commented:
/retest

cofyc commented:
@jsafrane https://github.com/kubernetes/kubernetes/compare/b976ebd56a07ebb009aeb8f14913e3472e5609ee...cofyc:noexec_evalsymlinks, FYI.

BTW, does `readlink` work?

msau42 commented on pkg/util/mount/mount_linux.go:
```diff
@@ -687,102 +684,137 @@ func doMakeRShared(path string, mountInfoFilename string) error {
-	evalSubPath, err := filepath.EvalSymlinks(subpath.Path)
-	if err != nil {
-		return "", fmt.Errorf("evalSymlinks %q failed: %v", subpath.Path, err)
+func safeOpenSubPath(mounter Interface, subpath Subpath) (int, error) {
```
Add comment this is also used by nsenter

jsafrane replied:
Comment added

msau42 commented on pkg/util/mount/mount_windows.go:
```diff
@@ -452,9 +450,23 @@ func (mounter *Mounter) GetFSGroup(pathname string) (int64, error) {
 }
 
+func (mounter *Mounter) GetMode(pathname string) (os.FileMode, error) {
+	info, err := os.Stat(pathname)
```
evalsymlinks?

jsafrane replied:
Symlinks are evaluated by the OS.

jsafrane replied:
Good point, I removed it from mount_linux

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -281,65 +276,164 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+func (mounter *NsenterMounter) ExistsPath(pathname string) (bool, error) {
+	// Resolve the symlinks but allow the target not to exist. EvalSymlinks
+	// would return an generic error when the target does not exist.
+	hostPath, err := mounter.ne.EvalSymlinks(pathname, false /* mustExist */)
```
Can you just do a `stat` on the host to check if it exists?

jsafrane replied:
How? `/bin/stat` returns exit code 1 for both permission denied and file not found errors.

msau42 commented on pkg/util/mount/mount.go:
```diff
@@ -84,16 +84,15 @@ type Interface interface {
+	// symlinks. Note that the function makes sure that it creates the directory
+	// somewhere under the base, nothing else. E.g. if the directory already
+	// exists, it may exist outside of the base due to symlinks.
+	SafeMakeDir(subdir string, base string, perm os.FileMode) error
```
Can you add a comment here this should be used if the path is inside the user's volume.  (to distinguish it from MakeDir)

jsafrane replied:
Comment added

msau42 commented on pkg/util/mount/nsenter_mount.go:
```diff
@@ -281,65 +276,164 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+
+	// Check that the bind-mount target is the same inode and device as the
+	// source that we keept open, i.e. we mounted the right thing.
+	var srcStat, dstStat unix.Stat_t
```
Can this section be put into a method for unit testing?

jsafrane replied:
I could, but how would I test it? I can't create mounts in unit test to check various scenarios.

msau42 replied:
It could at least be tested by using different files, not mounts.  I'm mainly concerned that e2es are not going to cover this path either.

jsafrane replied:
Ok, I added some unit tests that tries to find two directories on different filesystems. I don't like /dev/shm usage there, but it should not matter much in Kubernetes tests.

jsafrane replied:
Apparently /var/tmp is not writable in unit tests, I'll try to fix the heuristics.

msau42 reviewed (commented):
This can be squashed a bit. I just have some minor comments.

ncdc commented:
/uncc

jsafrane commented:
I squashed the tests into one commit, I could squash more, however I'd like the review chunks to be small, this PR is quite big to digest with one commit.

msau42 commented on pkg/volume/host_path/host_path.go:
```diff
@@ -350,7 +351,9 @@ type fileTypeChecker struct {
 func (ftc *fileTypeChecker) Exists() bool {
-	return ftc.mounter.ExistsPath(ftc.path)
+	exists, err := ftc.mounter.ExistsPath(ftc.path)
+	glog.Infof("path %s got %v, %v", ftc.path, exists, err)
```
Increase log level? This looks like debugging logs

jsafrane replied:
removed whole log line

jsafrane commented:
/test pull-kubernetes-local-e2e-containerized

jsafrane committed:
Refactor doBindSubPath into smaller functions:

- getSubpathBindTarget() computes final target of subpath bind-mount.
- prepareSubpathTarget() creates target for bind-mount.
- safeOpenSubPath() checks symlinks in Subpath and safely opens it.

jsafrane committed:
Allow EvalSymlinks target not to exist.

Various NsEnterMounter function need to resolve the part of the path that
exists and blindly add the part that doesn't.

jsafrane committed:
Enhance ExistsPath check

It should return error when the check fails (e.g. no permissions, symlink link
loop etc.)

jsafrane committed:
Change SafeMakeDir to resolve symlinks in mounter implementation

Kubelet should not resolve symlinks outside of mounter interface.
Only mounter interface knows, how to resolve them properly on the host.

As consequence, declaration of SafeMakeDir changes to simplify the
implementation:
from SafeMakeDir(fullPath string, base string, perm os.FileMode)
to   SafeMakeDir(subdirectoryInBase string, base string, perm os.FileMode)

jsafrane committed:
Created directories in /var/lib/kubelet directly.

jsafrane committed:
Pass Nsenter to NsenterMounter and NsenterWriter

So Nsenter is initialized only once and with the right parameters.

jsafrane committed:
Nsenter unit tests

msau42 commented on pkg/util/mount/nsenter_mount_test.go:
```diff
@@ -72,120 +75,668 @@ func TestParseFindMnt(t *testing.T) {
+		// Use /dev/shm as the last resort. It has different purpose than /tmp,
+		// but it's usually world-writable tmpfs and we clean up our mess
+		// quickly. Let's hope nobody notices.
+		"/dev/shm",
```
yeahhhhh I don't know about /dev/shm...  Maybe just /tmp or /var/tmp? 

You can still run the 2 "same fs" test cases on any system.  It's only the 3rd one that may not be able to run everywhere.

jsafrane replied:
Ha! I don't need two writable directories, one /tmp is enough. I can test with already existing file that's for sure on a different device, such as /proc/self/status - it guaranteed to exists, be readable and on a different "device" than /tmp.

jsafrane committed:
Split NsEnterMounter and Mounter implementation of doBindSubpath

nsenter implementation needs to mount different thing in the end and do
different checks on the result.

jsafrane committed:
Add GetMode to mounter interface.

Kubelet must not call os.Lstat on raw volume paths when it runs in a container.
Mounter knows where the file really is.

jsafrane commented:
/retest

msau42 commented:
/lgtm

jsafrane commented:
/assign @brendandburns @dchen1107 @lavalamp @smarterclayton @thockin
for approval

cc @liggitt, this is potentially another source of subpath regressions.

cofyc commented:
LGTM

cofyc commented on pkg/util/nsenter/nsenter.go:
```diff
@@ -139,11 +177,60 @@ func (ne *Nsenter) EvalSymlinks(pathname string) (string, error) {
+	return NewNsenter(rootfsPath, executor)
+}
+
+type fakeExec struct {
```
nit: Is it possible to use [`testing.FakeExec`](https://godoc.org/k8s.io/utils/exec/testing#FakeExec) in `nsenter_test.go`?

jsafrane commented:
/retest

liggitt commented:
> cc @liggitt, this is potentially another source of subpath regressions.

for which scenarios? does this include tests to cover those scenarios? also, do we now have CI running a containerized kubelet to exercise this code?

msau42 commented:
It's a regression for the scenario where:
* Contianierized kubelet
* Pod is using hostpath subpath
* Subpath path exists in the containerized kubelet image

To fix this, there is one new requirement in containerized kubelet deployment that /rootfs needs to be mounted rw

msau42 commented:
The current e2e (non containerized) has hostpath test cases but not with an existing path with the kubelet image

liggitt commented:
> To fix this, there is one new requirement in containerized kubelet deployment that /rootfs needs to be mounted rw

any new requirements should be called out in an action-required release note, right?

liggitt commented:
> The current e2e (non containerized) has hostpath test cases but not with an existing path with the kubelet image

if we support the containerized kubelet, it seems like we should exercise it in CI, especially given the complexity around how it has to be handled

thockin commented:
I'm very grumpy that pkg/util contains the word "kubelet" at all.  How can we fix this?  I will approve the PR since it fixes a bug and doesn't make things worse, but that's a genericity problem.

/approve

jsafrane commented on pkg/util/mount/mount.go:
```diff
@@ -84,16 +84,18 @@ type Interface interface {
+	// that's under user control. User must not be able to use symlinks to
+	// escape the volume to create directories somewhere else.
+	SafeMakeDir(subdir string, base string, perm os.FileMode) error
+	// Will operate in the host mount namespace if kubelet is running in a container.
```
Kubelet is the only user of this API and it makes the whole picture more visible. It would be harder to go through all the layers and see how they fit together if they contain just generic comments.

We could move the big picture to a separate file / documentation, however, these tend to rot.

I filled #64603 for that.

msau42 replied:
There is ongoing work to refactor the mount library to separate out k8s specific operations (like subpath processing), and general mounting/formatting utilities.  https://github.com/kubernetes/kubernetes/pull/68513

jsafrane commented:
/kind bug
/priority important-soon

jsafrane commented:
@saad-ali @childsb, please approve for 1.11

dims commented:
/status approved-for-milestone

munnerz commented:
@msau42 I'm running kubelet under rkt using the kubelet wrapper, with kube 1.11.0.

> To fix this, there is one new requirement in containerized kubelet deployment that /rootfs needs to be mounted rw

My systemd unit does not currently have rootfs mounted into the kubelet container, so I have attempted to add it in an attempt to take advantage of this patch. When doing so, kubelet then refuses to start with:

```
failed to run Kubelet: could not detect clock speed from output: ""
```

I'm wondering if I'm missing something obvious here? I am able to mount standard host path volumes fine, but when using the prometheus operator (which *enforces* a subPath be used), it will not work.

Is there any more detailed info on how I should configure this? I'd imagine (hope!) that tectonic doesn't suffer from this same problem, however I'm unsure where to dig into to find this out, and I'd prefer not to have to spin up an entire tectonic cluster just to find a couple of arguments! 😄

jsafrane commented:
@munnerz, list of directories that must be in the container with kubelet is here:

https://github.com/kubernetes/kubernetes/blob/7786bd8c9a99974e2cda31940dd4a1ef0a31c2e5/hack/local-up-cluster.sh#L825-L832

Please report if there is anything missing.

lichuqiang commented:
@jsafrane 
I ran into a problem when running the unit test in "nsenter_mount_test.go",
and got error like this:

```
--- FAIL: TestNsenterExistsFile (0.01s)
    nsenter_mount_test.go:343: Test "simple non-accessible file": expected error, got none
    nsenter_mount_test.go:347: Test "simple non-accessible file": expected return value false, got true
E0719 09:26:23.220671   67952 mount_linux.go:495] format of disk "/dev/foo" failed: type:("ext4") target:("/tmp/mount681850900") options:(["defaults"])error:(formatting failed)
E0719 09:26:23.221027   67952 mount_linux.go:529] Could not determine if disk "/dev/foo" is formatted (exit 4)
FAIL
FAIL    k8s.io/kubernetes/pkg/util/mount    0.077s
```

I switched to a non root user, and it got passed.
Is it required to manually switch to other users to run the case?
